src/main/java/org/skife/jdbi/v2/sqlobject/ResultReturnThing.java

```diff
@@ -102,11 +104,32 @@ else if (return_type.isInstanceOf(Iterator.class)) {
                 this.containerType = null;
             }
 
+            maxRowCount = fetchMaxRowCount(method);
+        }
+
+        private int fetchMaxRowCount(ResolvedMethod method){
+            if(method.getRawMember().isAnnotationPresent(MaxRowCountConstraint.class)){
+                MaxRowCountConstraint mrc = method.getRawMember().getAnnotation(MaxRowCountConstraint.class);
+                return mrc.value() > 0 ? mrc.value() : 0;
+            }
+            return 0;
         }
 
         @Override
         protected Object result(ResultBearing q, HandleDing baton)
         {
+            if (maxRowCount == 0) {
+                return returnFirst(q);
+            }
+
+            if (q.list(maxRowCount + 1).size() > maxRowCount) {
```

Comment:
This is very inefficient. You are executing the query and pulling down all data in all rows just to verify the row count. Better to use a `select count(*)` query instead and avoid all the network and disk i/o.